### DISCUSSION TITLE
Queue server returned error: 'Must call queue_traj_point service' (2)

### DISCUSSION THREAD

**Discussion by marip8** [gavanderhoorn/motoros2_fjt_pt_queue_proxy#2] [Q&A] [answered]
I'm testing this node at a deployment location today, and I ran into the following issue executing a trajectory that previously worked with the MotoROS2 `follow_joint_trajectory` action:

```
[WARN] [1709919117.174998813] [motoros2_fjt_pt_queue_proxy]: queue server returned error: 'Must call queue_traj_point service' (2)
[ERROR] [1709919117.175320074] [motoros2_fjt_pt_queue_proxy]: failed to queue pt 0, aborting goal (queue server reported: 2)
```

I have confirmed that the `queue_traj_point` service is being hosted by the MotoROS2 driver. I am also successfully able to call other MotoROS2 services (like `start_traj_mode`, and `read_single_io`, etc.), and I see the current position of the robot, so I'm fairly confident I can communicate to the driver.

Any ideas what is going on here? Am I missing some initialization step?

**Comment by gavanderhoorn** [comment, marked as answer]
Could you provide some more details on how you're using this (ie: control flow, services called, etc)?

As a general comment: MotoROS2 must already be in point-queueing mode (ie: someone / something must have called the `start_point_queue_mode` prior to the streamer node trying to enqueue points). The node in this repository does not do that.

The error you quote is returned [here](https://github.com/Yaskawa-Global/motoros2/blob/7a3019e4298def154af0233fcf39ec58fe4a265e/src/ActionServer_FJT.c#L259-L264), and it's raised when the motion mode isn't correct for the motion interface you're trying to use.

I'm not sure about the exact error message. I'm wondering why it mentions `queue_traj_point` instead of `start_point_queue_mode`. I'll log that as something to investigate.

**Reply by marip8** [reply]
> As a general comment: MotoROS2 must already be in point-queueing mode (ie: someone / something must have called the start_point_queue_mode prior to the streamer node trying to enqueue points).

I see. I didn't realize there was a separate service needed to put it into queuing mode. I had just been calling `start_traj_mode` before sending a trajectory to the proxy

**Reply by gavanderhoorn** [reply]
Well this node basically just calls [queue_traj_point](https://github.com/Yaskawa-Global/motoros2/blob/7a3019e4298def154af0233fcf39ec58fe4a265e/doc/ros_api.md#queue_traj_point) repeatedly for you, with your trajectory, so you don't have to.

It doesn't do much else at the moment. It's highly WIP.

The documentation on `queue_traj_point` does mention:

> The `start_point_queue_mode` service must have been called prior to attempting this service.

(that's not me telling you to RT*M, just drawing attention to the documentation)

**Comment by marip8** [comment]
We're getting another issue now about the driver failing to initialize the trajectory

```commandLine
[server-10]  [WARN] [1710368021.418699054] [fjt_proxy]: queue server returned error: 'Failed to initialize the streaming trajectory' (3)
```

We first checked the robot status topic which looked good, called the `start_point_queue_mode` service, heard and confirmed the servo drives were on, but the robot didn't move. We also confirmed that the robot start state and trajectory start state matched (started at the all-zeros state). The trajectory only has 5 points, each occurring 5 seconds after the previous with only a few degrees of motion per joint. Is this a trajectory problem of some sort, a weird proxy node problem, or a MotoROS2 problem?

**Reply by gavanderhoorn** [reply]
Hard to say, I'd need more info.

Could you run the [debug listener](https://github.com/Yaskawa-Global/motoros2/blob/5f8c80011c64dc9152a4c72b2eb138760a9d649e/doc/troubleshooting.md#debug-log-client) and then try to execute a trajectory?

Ideally start the listener before you start the controller. That way we can be sure you capture all relevant debug info.

If it turns out to be a MR2 problem, we'll open an issue there.

**Reply by marip8** [reply]
Sounds good. @bryanmqz can you test again with this debugger running?

**Reply by marip8** [reply]
Turns out it was our fault in the test trajectory. We were testing this node on a robot at SwRI for another robot elsewhere with a different configuration. First we didn't have the joint names correct for our SwRI robot. Then we were accidentally adding extra joints for a positioner that exists on our deployment robot and not on our SwRI robot. Once we fixed those issues, it worked fine.

That said, it would be nice to get more descriptive error messages about these issues from the driver. Even with the statements from the debug script, it wasn't entirely obvious what the error was. It would be nice if the driver would tell you 
- that you tried commanding more or fewer joints than it knows how to control
- that it doesn't recognize joint names in the trajectory

I'll open an issue on the MotoROS2 repo with some more details to track this

**Reply by gavanderhoorn** [reply]
I'll wait for you to post your issue, but IIRC, most of the things you mention are already checked by MotoROS2 (see [here](https://github.com/Yaskawa-Global/motoros2/blob/5f8c80011c64dc9152a4c72b2eb138760a9d649e/src/MotionControl.c#L623-L745) fi).

The -- *very much a WIP* -- script in this repository does not. It's just a quick example of how someone might interact with the `queue_traj_point` service.

But let's see how we can improve the queuing interface.

**Reply by gavanderhoorn** [reply]
@marip8: just a friendly ping. I don't believe you've posted to the MotoROS2 issue tracker yet?

**Reply by marip8** [reply]
See [here](https://github.com/Yaskawa-Global/motoros2/discussions/230)
